File: frontend/src/api/research_agent.ts
import { API_BASE_URL } from '../lib/api'
import { getStoredToken } from './auth'

export type ResearchDigestPaper = {
  arxiv_id: string
  title: string
  authors: string[]
  abstract: string
  published_date: string
  categories: string[]
  pdf_url: string
  relevance_score: number
}

export type ResearchDigest = {
  summary: string
  key_findings: Array<{ topic: string; finding: string; evidence_papers: string[] }>
  methodologies: Array<{ name: string; frequency: number; papers: string[] }>
  limitations: string[]
  trends: Array<{ trend: string; direction: 'increasing' | 'decreasing' | 'stable'; recent_papers: string[] }>
  total_papers_reviewed: number
  papers_cited: ResearchDigestPaper[]
  search_duration_seconds: number
}

export type ResearchDigestResponse = {
  session_id: string
  chat_id: string
  user_message_id: string
  assistant_message_id: string
  query: string
  digest: ResearchDigest
  search_duration_seconds: number
  papers_found: number
}

function authJsonHeaders(): HeadersInit {
  const token = getStoredToken()
  return {
    'Content-Type': 'application/json',
    ...(token ? { Authorization: `Bearer ${token}` } : {}),
  }
}

function formatErrorDetail(err: unknown): string | null {
  if (!err || typeof err !== 'object') return null
  const detail = (err as { detail?: unknown }).detail
  if (typeof detail === 'string') return detail
  if (Array.isArray(detail)) {
    return detail
      .map((item) => {
        if (!item || typeof item !== 'object') return String(item)
        const msg = (item as { msg?: unknown }).msg
        const loc = (item as { loc?: unknown }).loc
        const locText = Array.isArray(loc) ? loc.join('.') : ''
        return `${locText ? `${locText}: ` : ''}${typeof msg === 'string' ? msg : JSON.stringify(item)}`
      })
      .join('; ')
  }
  if (detail !== undefined) return JSON.stringify(detail)
  return null
}

export async function runResearchQuery(payload: {
  query: string
  chat_id?: string
  depth?: 'quick' | 'balanced' | 'deep'
  max_papers?: number
}): Promise<ResearchDigestResponse> {
  const response = await fetch(`${API_BASE_URL}/api/research-agent/research`, {
    method: 'POST',
    headers: authJsonHeaders(),
    body: JSON.stringify({
      query: payload.query,
      chat_id: payload.chat_id,
      depth: payload.depth ?? 'balanced',
      max_papers: payload.max_papers ?? 20,
    }),
  })

  if (!response.ok) {
    let detail = `Failed to run research: ${response.status}`
    try {
      const err = (await response.json()) as unknown
      const parsed = formatErrorDetail(err)
      if (parsed) detail = parsed
    } catch {
      // ignore parse errors
    }
    throw new Error(detail)
  }

  return (await response.json()) as ResearchDigestResponse
}

File: backend/app/schemas/research_agent.py
"""
Schemas for Research Digest Agent.
"""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class ResearchPaperRef(BaseModel):
    """Reference to a research paper."""
    arxiv_id: str
    title: str
    authors: list[str]
    abstract: str
    published_date: str
    categories: list[str]
    pdf_url: str
    relevance_score: float
    inclusion_reason: str | None = None

    model_config = {"from_attributes": True}


class ResearchDigestKeyFinding(BaseModel):
    """A key finding extracted from research."""
    topic: str
    finding: str
    evidence_papers: list[str]  # arxiv_ids


class ResearchDigestMethodology(BaseModel):
    """Common methodologies found in research."""
    name: str
    frequency: int
    papers: list[str]  # arxiv_ids


class ResearchDigestTrend(BaseModel):
    """Research trends identified."""
    trend: str
    direction: str  # increasing, decreasing, stable
    recent_papers: list[str]  # arxiv_ids


class ResearchDigestFull(BaseModel):
    """Full structured research digest."""
    summary: str
    key_findings: list[ResearchDigestKeyFinding]
    methodologies: list[ResearchDigestMethodology]
    limitations: list[str]
    trends: list[ResearchDigestTrend]
    total_papers_reviewed: int
    papers_cited: list[ResearchPaperRef]
    search_duration_seconds: int


class ResearchSessionRead(BaseModel):
    """Read model for research session."""
    id: uuid.UUID
    chat_id: uuid.UUID
    user_id: uuid.UUID
    research_query: str
    status: str
    papers_found: int
    digest_summary: str | None = None
    digest_full: dict[str, Any] | None = None
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}


class ResearchQueryRequest(BaseModel):
    """Request to start research on a topic."""
    chat_id: uuid.UUID | None = None
    query: str = Field(min_length=5, max_length=2000)
    max_papers: int = Field(default=10, ge=1, le=50)
    depth: str = Field(default="balanced", pattern="^(quick|balanced|deep)$")  # quick, balanced, deep


class ResearchDigestStreamEvent(BaseModel):
    """Event streamed during research."""
    event_type: str  # searching, found_paper, analyzing, generating_digest, completed, error
    data: dict[str, Any]
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class ResearchDigestResponse(BaseModel):
    """Final research digest response."""
    session_id: uuid.UUID
    chat_id: uuid.UUID
    user_message_id: uuid.UUID
    assistant_message_id: uuid.UUID
    query: str
    digest: ResearchDigestFull
    search_duration_seconds: int
    papers_found: int
